### PR TITLE
OOP school library: set up associations

### PR DESCRIPTION
**Pull request by ahadb802**
# OOP school library: set up associations

### Description
In this project, I have finished creating the remaining classes for our school library and created the associations between them.

### Project requirements
- Create a class `Classroom` with the following:
  - `@label` instance variable, should be initialized in the constructor.
  - Setter and getter for `@label` (remember about `attr_accessor`).
- Create the `has-many`/`belongs-to` relationship between `Classroom` and `Student`. The following should be implemented:
  - Create the `has-many` side (a classroom has many students).
  - Create the `belongs-to` side (a student belongs to a classroom).
  - Make sure that when adding a student to a classroom it also sets the classroom for the student.
  - Make sure that when setting the classroom for a student it also adds it to the classrooms' students.
- Create a class `Book` with the following:
  - `@title` and `@author` instance variables, should be initialized in the constructor.
  - Setters and getters for instance variables (remember about `attr_accessor`).
- Create a class `Rental` with the following:
  - `@date` instance variable, should be initialized in the constructor.
  - Setter and getter for `@date` (remember about `attr_accessor`).
- Create the `many-to-many` (also `has-many-through`) relationship between `Person` and `Book` using the intermediate class `Rental`. The following should be implemented:
  - Create the `has-many` side of `Book` and `Rental` (a book has many rentals).
  - Create the `belongs-to` side of `Rental` and `Book` (a rental belongs to a book).
  - Create the `has-many` side of `Person` and `Rental` (a person has many rentals).
  - Create the `belongs-to` side of `Rental` and `Person` (a rental belongs to a person).
  - Modify the constructor of `Rental` so `Book` and `Person` are set in it.

### CODE REVIEW

File: src/student.rb [hunk 1]
```diff
@@ -1,7 +1,11 @@
+require_relative './person'
 class Student < Person
-  def initialize(age, classroom, name = 'unknown', parent_permission: true)
-    super(age, name, parent_permission)
+  attr_accessor :classroom
+
+  def initialize(age, name = 'Unknown', parent_permission: true, classroom: nil)
+    super(age, name, parent_permission: parent_permission)
     @classroom = classroom
+    classroom&.add_student(self)
```
Comment (mikethreels) [inline src/student.rb]: - [ ] this part of code should be placed in a separate method

File: src/books.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+class Book
+  # ...
+
+  attr_reader :rentals
+
+  def initialize(_title, _author)
+    # ...
+    @rentals = []
+  end
```
Comment (mikethreels) [inline src/books.rb]: - [ ] the @title and @author instance variables should be initialized in the constructor. please find a link to the example about constructors in a previous lesson https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/class_and_object_for_irb.md#constructor-and-instance-vars
- [ ] there should be a setter and getter created for these instance variables as well.

File: src/books.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+class Book
+  # ...
+
+  attr_reader :rentals
+
+  def initialize(_title, _author)
+    # ...
+    @rentals = []
```
Comment (mikethreels) [inline src/books.rb:10]: - [ ] A method for adding rentals should be present. please find a link to the example in a previous lesson: https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#many-to-many-relationship

File: src/person.rb [hunk 4]
```diff
@@ -1,16 +1,28 @@
...
-    @id = rand(1..1000)
-    @name = name
-    @age = age
-    @parent_permission = parent_permission
+require_relative './nameable'
+
+class Person < Nameable
+  # ...
+
+  attr_reader :rentals
```
Comment (mikethreels) [inline src/person.rb]: - [ ] Getters and Setters for @rentals, @name, @age should be present

**Review by mikethreels — CHANGES_REQUESTED** [review]
Hi Ahad, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- linters are passing

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by victorokoroji — APPROVED** [review]
Hi @ahadb802, 

# ✔️   Status: APPROVED 🏆 🎉 🎊 

You have done a great job on this project, keep it up 👍 🥇. Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
congratulations!  🎉💯🌟 

## Top highlight 🎯

- The Pull Request is following the Git Flow ✅ 
- The PR title and description are professional and descriptive ✅ 
- The commit messages are meaningful ✅ 



## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**